### DISCUSSION TITLE
OpenTelemetry: Improve Trace-Propagation

### DISCUSSION THREAD

**Discussion by lforst** [vercel/next.js#64862] [Ideas]
### Goals

The Next.js OpenTelemetry instrumentation should be decisive about trace propagation since the current implementation makes it difficult to use OpenTelemetry instrumentation.

### Background

Currently, Next.js propagates OpenTelemetry trace data for incoming requests, but not for outgoing requests. This makes Next.js very tricky to use with existing official OpenTelemetry instrumentation (like the OTEL [http instrumentation](https://www.npmjs.com/package/@opentelemetry/instrumentation-http)) which generally instruments both, incoming, and outgoing requests. To be able to use the OTEL instrumentation, you need to either, very expensivley, stomp all spans created by Next.js, or weirdly override the OpenTelemetry instrumentation.

Since it is very common to use the OpenTelemetry HTTP instrumentation, Next.js is probably doing more harm in this situation than good.

### Proposal

**My proposed solution for this is pretty straightforward:**

- [ ] **Next.js should not concern itself with trace-propagation**
- [ ] **EXCEPT for when propagation requires framework-level capabilities (for example, propagating traces to the client for SSR: https://github.com/vercel/next.js/pull/64256)**

That way, people can, without any complications, implement their own propagation instrumentation logic, without Next.js interfering.

An alternative solution would be for Next.js to instrument and propagate _everything_ (incoming reqs, outgoing http reqs, outgoing fetch reqs, client propagation), but then Next.js has to make guarantees about it.

**Comment by dvoytenko** [comment]
It might be better to keep outbound instrumentation outside of Next. This is a path we took for `@vercel/otel` [here](https://github.com/vercel/otel/blob/main/packages/otel/src/instrumentations/fetch.ts).

**Reply by lforst** [reply]
I agree, hence why I originally suggested Next.js should probably not do any propagation. Having thought about it, propagating incoming traces is probably still valuable. 🤔

However, I think Next.js should still propagate in situations where it is impossible for libraries to automatically propagate, like for SSR.

So I think should this PR (https://github.com/vercel/next.js/pull/64256) ever be merged, we can probably close this discussion.
